src/test/ui/consts/control-flow/drop.rs

```diff
@@ -0,0 +1,43 @@
+// Ensure that a value is never dropped inside a `const`.
+
+#![feature(const_if_match)]
+
+// OK: `x` is always moved into the final value and is not dropped inside the initializer.
+const _: Option<Vec<i32>> = {
+    let y: Option<Vec<i32>> = None;
+    let x = Some(Vec::new());
+
+    if true {
+        x
+    } else {
+        x
+    }
+};
+
+// BAD: `x` is *not* always moved into the final value may be dropped inside the initializer.
+const _: Option<Vec<i32>> = {
+    let y: Option<Vec<i32>> = None;
+    let x = Some(Vec::new());
+    //~^ ERROR destructors cannot be evaluated at compile-time
+
+    if true {
+        x
+    } else {
+        y
+    }
+};
+
+// BUG?: even though we move the `Vec` out of `x`, we don't mark `x` as no longer `NeedsDrop`.
```

Comment:
do you want to open an issue about it or are you addressing it already?

Reply:
We reject this for the same reason we reject:

```rust
const fn foo() -> Vec<i32> {
    let vec_and_bool = (Vec::new(), true);
    vec_and_bool.0
}
```

My thought was that we could special case that a move out of the only field of a struct or enum variant also clears `Drop` for the container. We may want to defer this to a follow-up PR? I'm not quite sure what all the implications of this are.

Reply:
Oh yea, please defer. Being overly restrictive at the start is totally the right way

Reply:
I added the reduced version as a failing test case.